rollup/internal/controller/relayer/l2_relayer_sanity.go

```diff
@@ -0,0 +1,438 @@
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get last batch by hash %s: %w", info.LastBatchHash.Hex(), err)
+	}
+
+	// Get all batches in the range (parent+1 to last)
+	firstBatchIndex := parentBatch.Index + 1
+	lastBatchIndex := lastBatch.Index
+
+	// Check if the range is valid
+	if firstBatchIndex > lastBatchIndex {
+		return nil, nil, nil, fmt.Errorf("no batches found in range: first index %d, last index %d", firstBatchIndex, lastBatchIndex)
+	}
+
+	var batchesToValidate []*dbBatchWithChunks
+	for batchIndex := firstBatchIndex; batchIndex <= lastBatchIndex; batchIndex++ {
+		dbBatch, err := r.batchOrm.GetBatchByIndex(r.ctx, batchIndex)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("failed to get batch by index %d: %w", batchIndex, err)
+		}
+
+		// Get chunks for this batch
+		dbChunks, err := r.chunkOrm.GetChunksInRange(r.ctx, dbBatch.StartChunkIndex, dbBatch.EndChunkIndex)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("failed to get chunks for batch %d: %w", batchIndex, err)
+		}
+
+		batchesToValidate = append(batchesToValidate, &dbBatchWithChunks{
+			Batch:  dbBatch,
+			Chunks: dbChunks,
+		})
+	}
+
+	// Get first batch for return
+	firstBatch := batchesToValidate[0].Batch
+
+	return batchesToValidate, firstBatch, lastBatch, nil
+}
+
+// validateCalldataAgainstDatabase validates calldata parameters against database records
+func (r *Layer2Relayer) validateCalldataAgainstDatabase(info *CalldataInfo, firstBatch, lastBatch *orm.Batch) error {
+	// Validate codec version
+	if info.Version != uint8(firstBatch.CodecVersion) {
+		return fmt.Errorf("version mismatch: calldata=%d, db=%d", info.Version, firstBatch.CodecVersion)
+	}
+
+	// Validate parent batch hash
+	if info.ParentBatchHash != common.HexToHash(firstBatch.ParentBatchHash) {
+		return fmt.Errorf("parentBatchHash mismatch: calldata=%s, db=%s", info.ParentBatchHash.Hex(), firstBatch.ParentBatchHash)
+	}
+
+	// Validate last batch hash
+	if info.LastBatchHash != common.HexToHash(lastBatch.Hash) {
+		return fmt.Errorf("lastBatchHash mismatch: calldata=%s, db=%s", info.LastBatchHash.Hex(), lastBatch.Hash)
+	}
+
+	return nil
+}
+
+// validateDatabaseConsistency performs comprehensive validation of database records
+func (r *Layer2Relayer) validateDatabaseConsistency(batchesToValidate []*dbBatchWithChunks) error {
+	if len(batchesToValidate) == 0 {
+		return fmt.Errorf("no batches to validate")
+	}
+
+	// Get previous chunk for continuity check
+	firstChunk := batchesToValidate[0].Chunks[0]
+	if firstChunk.Index == 0 {
+		return fmt.Errorf("genesis chunk should not be in normal batch submission flow, chunk index: %d", firstChunk.Index)
+	}
+
+	prevChunk, err := r.chunkOrm.GetChunkByIndex(r.ctx, firstChunk.Index-1)
+	if err != nil {
+		return fmt.Errorf("failed to get previous chunk %d for continuity check: %w", firstChunk.Index-1, err)
+	}
+
+	// Validate codec version consistency across all batches
```

Comment:
update comment to contain all checks or just remove

Reply:
The comment is fixed in https://github.com/scroll-tech/scroll/pull/1714/commits/be22866f4290b185574142c766d10036e45a7df3.